game/src/sandbox/dashboards/commuter.rs

```diff
@@ -19,7 +19,8 @@ pub struct CommuterPatterns {
     bldg_to_block: HashMap<BuildingID, BlockID>,
     border_to_block: HashMap<IntersectionID, BlockID>,
     blocks: Vec<Block>,
-    selected: Option<CurrentBlock>,
+    block_drawables: HashMap<(BlockID, BlockDrawableType), Drawable>,
```

Comment:
My approach was basically:

- compute and cache the required drawables in `event`
- clear the cache whenever necessary 🤞 
- `draw` just pulls out the necessary cached drawables and explodes if they're not there

I'm very open to feedback on this. I don't think I really grok the graphics cycle yet.

Reply:
The graphics cycle works like this:

- an event happens, like a keypress, mouse movement, or time passing (when running the simulation or animating something), `event()` gets called
- after some number of events or delay (determined by `winit`), `draw()` gets called

My philosophy on performance is to prioritize keeping the code as simple as possible, only doing something more complicated when there's noticeable lag in debug mode. The concern here is that when we're locked, we don't want to redo work in every single `draw` to figure out what block we're moused over, count everything, and upload the geometry with the text label and white outline. One of the earlier versions did that, and it was fixed [here](https://github.com/dabreegster/abstreet/commit/68561170bd733083af250f21eb282cd0b6e76361).

I think an entire map of `block_drawables` is overkill. We just want to cache one state at any time. It's a bit confusing with all the locked and unlocked cases. The expanded form:

```
enum CurrentBlock {
  NothingSelected,
  Unlocked(BlockID, Drawable),   // hovering on something before clicking
  LockedWithoutComparison(BlockID, Drawable),    // outline and in/out icon
  LockedWithComparison {
    locked: BlockID,
    compare_to: BlockID,
    draw: Drawable,   // both for the locked and comparison block
}
```

We want to maintain this in `event` when `redo_mouseover` says the mouse has moved. Maybe this expanded enum is too verbose, we could also try something similar to the current:

```
struct CurrentBlock {
  id: BlockID,
  draw: Drawable,  // all the buildings and outline
  locked: Option<Drawable>,  // the in/out icons
  compare_to: Option<(BlockID, Drawable)>,
}

We'd `g.redraw()` up to 3 `Drawables` with this representation. That's also fine from a perf perspective; you don't want to tell the GPU to redraw 1000's of things every frame, so batching is helpful there. As you move between `LockedWithComparison` states, you could either keep two of the old Drawables for the locked block, or just recompute them again. Likely both options are fine, just a matter of code organization.

Let me know if this doesn't make sense.

Reply:
Thanks for the feedback!

One thing that lead to my approach was the desire to separate computing state from generating the drawables, but now that we're caching the drawables in the CurrentBlock:: enum, it seems I'll have to sprinkle around the generation of drawables along with the setting of state.

LMK if you see a way to avoid that, but otherwise, I'll just abandon that particular design goal for now.

Reply:
It would be nice to separate the two concerns a bit. What if we were closer to 

```
enum CurrentBlock {
enum CurrentBlock {
  NothingSelected,
  Unlocked(BlockID),
  LockedWithoutComparison(BlockID),
  LockedWithComparison(BlockID, BlockID),
}
```

And in `event`, we first just focus on updating this. If there was a change, then compute everything to draw and do `composite.replace` in a separate function?